tuf/api/metadata.py

```diff
@@ -1,3 +1,38 @@
+"""TUF role metadata model.
+
+This module provides container classes for TUF role metadata, including methods
+to read/serialize/write from and to JSON, perform TUF-compliant metadata
+updates, and create and verify signatures.
+
+TODO:
+
+ * Add docstrings
+
+ * Finalize/Document Verify/Sign functions (I am not fully sure about expected
+   behavior)
+
+ * Validation (some thoughts ...)
+   - Avoid schema, see secure-systems-lab/securesystemslib#183
+   - Provide methods to validate JSON representation (at user boundary)
+   - Fail on bad json metadata in read_from_json method
+   - Be lenient on bad/invalid metadata objects in memory, they might be
+     work in progress. E.g. it might be convenient to create empty metadata
+     and assign attributes later on.
+   - Fail on bad json metadata in write_to_json method, but with option to
+     disable check as there might be a justified reason to write WIP
+     metadata to json.
+
+ * It might be nice to have short-cuts on the Metadata class to methods and
+   attributes of the contained Signed object. If we do this, we should only do
+   it on common methods/attributes (e.g. version, bump_version, expires,
+   bump_expiration)
```

Comment:
I'm undecided on this. I'm leaning in favour of not mapping attributes of the Signed object into the metadata class, because something like `timestamp.signed.expires` better matches the metadata format in the specification and that's what we set out to achieve with this re-working effort.

Reply:
Yeah, I'm a bit torn too. I think as long as it is clear that the `expires`, `version`, etc. are only convenience short-cuts/aliases to those attributes defined on the inner metadata it is okay. 

Still, it might be confusing to have aliases on the common attributes only, but not on those specific to any of timestamp, snapshot, targets... and adding conditional shortcuts to specific fields too might make the short-cutting implementation to complicated.

So if we agree that the typing effort (as in amount of used characters) of e.g. `timestamp.signed.expires` is still reasonable, we can go without aliases for the time being.

Besides, it might make sense to adopt a workflow, where we only work on Metadata objects when needed, and otherwise work directly on the inner metadata, e.g.:
```python

timestamp_md = Timestamp.load_from_json("timestamp.json")
timestamp_md.verify(keys)

# Discard container once existing signatures are verified and
# do all the heavy lifting directly on timestamp
timestamp = timestamp_md.signed
# timestamp.foo()
# timestamp.bar()
# ...

# Re-wrap timestamp in Metadata when signing is needed,
timestamp_md = Metadata(signed=timestamp)
timestamp_md.sign(keys)
timestamp_md.write_to_json()
```

Reply:
> So if we agree that the typing effort (as in amount of used characters) of e.g. `timestamp.signed.expires` is still reasonable, we can go without aliases for the time being.

I vote we go without aliases to start. In terms of API stability, it's easier to add them later if they are needed than it is to remove them.

> Besides, it might make sense to adopt a workflow, where we only work on Metadata objects when needed, and otherwise work directly on the inner metadata, e.g.:
> 
> ```python
> timestamp_md = Timestamp.load_from_json("timestamp.json")
> timestamp_md.verify(keys)
> 
> # Discard container once existing signatures are verified and
> # do all the heavy lifting directly on timestamp
> timestamp = timestamp_md.signed
> # timestamp.foo()
> # timestamp.bar()
> # ...
> 
> # Re-wrap timestamp in Metadata when signing is needed,
> timestamp_md = Metadata(signed=timestamp)
> timestamp_md.sign(keys)
> timestamp_md.write_to_json()
> ```

Nice idea. We should certainly give ourselves time to experiment with this pattern and see how it feels in use.

Reply:
> I vote we go without aliases to start. In terms of API stability, it's easier to add them later if they are needed than it is to remove them.

Agreed.

> Nice idea. We should certainly give ourselves time to experiment with this pattern and see how it feels in use.

Forgot to mention, this is what we e.g. do in [layout in-toto's verification routine](https://github.com/in-toto/in-toto/blob/a621c3bf3cbe1dd442b2388c685b3130f032430b/in_toto/verifylib.py#L1462-L1465).

Reply:
I like this idea. Feels a lot better than trying to update signatures in place...

Reply:
Sorry for the late response -- I agree with @joshuagl here.